File: apps/webflux/chatting-service/src/main/java/com/booster/chattingservice/config/RedisConfig.java
package com.booster.chattingservice.config;

import com.booster.chattingservice.dto.ChatMessage;
import com.booster.chattingservice.service.ChatServiceImpl;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.redis.connection.ReactiveRedisConnectionFactory;
import org.springframework.data.redis.listener.ChannelTopic;
import org.springframework.data.redis.listener.ReactiveRedisMessageListenerContainer;
import tools.jackson.databind.ObjectMapper;

@Slf4j
@Configuration
@RequiredArgsConstructor
public class RedisConfig {

    private final ChatServiceImpl chatService; // 구체 클래스 주입 (broadcastToLocalUsers 호출용)
    private final ObjectMapper objectMapper;

    @Bean
    public ReactiveRedisMessageListenerContainer redisMessageListenerContainer(
            ReactiveRedisConnectionFactory factory) {

        ReactiveRedisMessageListenerContainer container = new ReactiveRedisMessageListenerContainer(factory);

        // "chat.public" 토픽을 구독
        container.receive(ChannelTopic.of("chat.public"))
                .map(record -> record.getMessage()) // 메시지 내용(String) 추출
                .doOnNext(json -> {
                    try {
                        // JSON -> Record 변환
                        ChatMessage message = objectMapper.readValue(json, ChatMessage.class);
                        // 서비스 계층으로 전달 -> 로컬 유저들에게 발송
                        chatService.broadcastToLocalUsers(message);
                    } catch (Exception e) {
                        log.error("Failed to process Redis message", e);
                    }
                })
                .subscribe(); // 중요: 스트림 구독 시작

        return container;
    }
}

File: apps/kotlin/chatting-service/src/main/kotlin/com/booster/kotlin/chattingservice/config/RedisConfig.kt
package com.booster.kotlin.chattingservice.config

import com.booster.kotlin.core.logger
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.data.redis.connection.ReactiveRedisConnectionFactory
import org.springframework.data.redis.listener.ReactiveRedisMessageListenerContainer

/**
 * Redis 설정.
 *
 * ReactiveRedisMessageListenerContainer만 Bean으로 등록.
 * 채널 구독은 ChatService가 방(room) 단위로 동적 관리:
 *   - 첫 유저 입장 시 chat.room.{roomId} 채널 구독
 *   - 마지막 유저 퇴장 시 채널 구독 해제
 *
 * 이전 구조 (chat.public 단일 채널):
 *   모든 서버가 모든 메시지를 수신 → 불필요한 처리 발생
 *
 * 변경 후 구조 (chat.room.{roomId} 방별 채널):
 *   해당 방에 유저가 있는 서버만 구독 → 불필요한 수신 제거
 */
@Configuration
@ConditionalOnProperty(name = ["chat.redis.enabled"], havingValue = "true", matchIfMissing = true)
class RedisConfig {

    private val log = logger()

    @Bean
    fun redisMessageListenerContainer(
        factory: ReactiveRedisConnectionFactory
    ): ReactiveRedisMessageListenerContainer {
        log.info("[REDIS] ReactiveRedisMessageListenerContainer 생성 완료")
        return ReactiveRedisMessageListenerContainer(factory)
    }
}
